### ISSUE TITLE
Crashlytics `upload-symbols` crashing when launched via `run` script in Xcode Build Phase

### ISSUE THREAD

**Issue by elincoln** [firebase/firebase-ios-sdk#7940]
### [REQUIRED] Step 1: Describe your environment

  * Xcode version: 12.4 (12D4e)
  * Firebase SDK version: 7.11.0
  * Installation method: `Swift Package Manager` (select one)
  * Firebase Component: Crashlytics

### [REQUIRED] Step 2: Describe the problem

`upload-symbols` via the `run` script is **sometimes** crashing in the background/async (after passing the `--validate` synchronous step)

Here's the synchronous portion of `run` in the Xcode build log:

> upload-symbols 3.7 build 26
> Running upload-symbols in Build Phase mode
> Validating build environment for Crashlytics...
> 
> Arguments:
> Debug mode enabled
> Google App ID: `<<REDACTED>>`
> Platform: ios
> DSYM Paths: ["/Users/ethan/Library/Developer/Xcode/DerivedData/Unearth-gbsbyrknfhezwchjzlpwdwfzssic/Build/Products/Debug-iphonesimulator/Unearth.app.dSYM"]
> 
> Validation succeeded. Exiting because upload-symbols was run in validation mode
> 

Here's the crash and diagnostic reports it generates: 
[Ethan-upload-symbols-crash-and-diag-reports.zip](https://github.com/firebase/firebase-ios-sdk/files/6361022/Ethan-upload-symbols-crash-and-diag-reports.zip)

Here's a Console log snippet right before it crashes:
```
default	13:13:30.456672-0700	upload-symbols	Read (Async) options: 80001 -- URL: <private> -- purposeID: A563325A-9BF6-4684-97F7-AF86FA5DE3E2 -- claimID: 4C068151-05BA-4338-94F0-15DD740CB769
default	13:13:30.457174-0700	upload-symbols	Claim 4C068151-05BA-4338-94F0-15DD740CB769 granted in client
default	13:13:30.457194-0700	upload-symbols	Claim 4C068151-05BA-4338-94F0-15DD740CB769 invoked in client
default	13:13:30.457531-0700	upload-symbols	_LSBundleCopyOrCheckNode: cached node not found, _LSBundleCreateNode for bundleID 36276 returned -43
default	13:13:30.458714-0700	upload-symbols	Read (Async) options: 80001 -- URL: <private> -- purposeID: D1E66B71-BB82-459D-A502-0E5923958F2E -- claimID: 846386DC-7C12-46FC-AD31-37DDCAB37C33
default	13:13:30.459217-0700	upload-symbols	Claim 846386DC-7C12-46FC-AD31-37DDCAB37C33 granted in client
default	13:13:30.459249-0700	upload-symbols	Claim 846386DC-7C12-46FC-AD31-37DDCAB37C33 invoked in client
default	13:13:30.460030-0700	upload-symbols	Read (Async) options: 80001 -- URL: <private> -- purposeID: A6ED3869-BF8B-4C18-BBF6-710C651414CC -- claimID: E0DB0B8A-7BCB-407B-AAF8-6F6932232965
default	13:13:30.460536-0700	upload-symbols	Claim E0DB0B8A-7BCB-407B-AAF8-6F6932232965 granted in client
default	13:13:30.460602-0700	upload-symbols	Claim E0DB0B8A-7BCB-407B-AAF8-6F6932232965 invoked in client
default	13:13:30.461095-0700	upload-symbols	_LSBundleCopyOrCheckNode: cached node not found, _LSBundleCreateNode for bundleID 36344 returned -43
default	13:13:32.056185-0700	kernel	apfs_pagein:9334: ino 30420853/30420853 pl_offset 0 f_offset 2940928 size 4096 file size 0 error 22 (upload-symbols)
default	13:13:32.093929-0700	ReportCrash	Parsing corpse data for process upload-symbols [pid 92827]
default	13:13:34.510701-0700	ReportCrash	osa_update: Pid 92827 'upload-symbols' CORPSE: Parsing KCData
```

When I manually run ` upload-symbols -d -ai "<<REDACTED>>" -p "ios" Unearth.app.dSYM`, it uploads fine.

I even copy-pasted the entire, massive environment variable export produced by Xcode, and ran `run --debug` it uploads fine!

Struggling to figure out what why this is failing seemingly at random when launched via Xcode Build Phase.

Here's my build-phase settings: 
<img width="855" alt="image" src="https://user-images.githubusercontent.com/1448144/115781456-4046a100-a36f-11eb-9cbf-f4cb12f918f7.png">

**Comment by rizafran** [comment]
Thanks for reporting @elincoln. It looks like your issue is related here #7634.

**Comment by samedson** [comment]
@elincoln sorry this is happening. We recently added code to validate dSYMs that may be the culprit. Could you try adding the following flag to `run` to see if the crashes stop happening `--skip-dsym-validation` ?

**Comment by elincoln** [comment]
Hi @samedson, 
**TLDR:** that fixed the crash but `upload-symbols` is now uploading the old/stale dSYM file for local incremental builds, and not the new one. 

**Details:**
Adding `--skip-dsym-validation` does prevent the crash, but your mention of a race condition made me look a little closer: while it no longer crashes on incremental builds, it instead uploads a stale dSYM file.

Before performing an incremental build, i checked the dSYM UUID w/ `dwarfdump -u Unearth.app.dSYM` and it was `5C3CA724-...`

I modified your `run` script to dump the output of `upload-symbols` to a file instead of /dev/null, and pass it `--skip-dsym-validation` and `--debug`, I get the following output:

[upload-symbols-logs.txt](https://github.com/firebase/firebase-ios-sdk/files/6367733/upload-symbols-logs.txt)
The key line in there is `Uploading cSYM for uuid 5c3ca724...`

and after the build completed, i checked the new dSYM UUID, finding it to be `18499242...` (so it clearly uploaded the wrong one)

However, if i delete the dSYM before doing an incremental, `upload-symbols` spams "Unable to get file attributes for dSYM file at path..." until the dSYM appears, and then it uploads the correct, new one.

Hope this helps you find a reliable resolution.

**Comment by samedson** [comment]
Dang that is indeed problematic - thank you for this information. 

Were you able to try adding the `${DWARF_DSYM_FOLDER_PATH}/${DWARF_DSYM_FILE_NAME}/Contents/Resources/DWARF/${TARGET_NAME}` to the "Input Files" section of the run / upload-symbols script in Build Phases?

**Comment by elincoln** [comment]
> Dang that is indeed problematic - thank you for this information.
> 
> Were you able to try adding the `${DWARF_DSYM_FOLDER_PATH}/${DWARF_DSYM_FILE_NAME}/Contents/Resources/DWARF/${TARGET_NAME}` to the "Input Files" section of the run / upload-symbols script in Build Phases?

🤦  Adding this fixed my "uploading stale symbols" issue.  It even seems to have resolved the crash; WITHOUT needing to include the `--skip-dsym-validation`

My apologies for poor reading comprehension on my part -- I even tried using that setting (incorrectly) and somehow overlooked that part of your post! 

For future reference, I'm guessing that if my build produced multiple dSYM files that I would need to add them to the input files section?

Anyway, consider MY issue resolved.  (But it sounds like you have a fun problem to solve, making this easier for more complex builds than mine 😆)

**Comment by samedson** [comment]
All good - thank you for reaching out and carefully describing the issue! Very glad it's resolved on your end. Yes for additional dSYMs you'll want to put them into the Input Files. You may also need to add an additional call to upload-symbols, eg. `upload-symbols --build-phase -- <path-to-dsym>`.

Haha but yeah... we fixed a similar issue to this a while back with Xcode 10 when they introduced async builds (meaning the run script could be run *while* Xcode generated the dSYM). That's what you see when upload-symbols prints "waiting..." log messages. Now we need to figure out why Xcode isn't deleting the dSYM right away.

We'll be doing some work on this in the future to improve both multiple dSYMs and Xcode build race conditions.
